kernel: some cleanup and fixes

nashif opened the pull request:
First step to reduce complixity of some files under kernel/ which have been growing in size and scope, particularly sched.c and thread.c
No functional changes, just code, header and kconfig reorg and cleanup.

nashif committed:
kernel: split cpu_mask handling into own file

In an effort to cleanup sched.c, move sections of code that can be
compiled in based on options into own files. CPU mask here is managed by
a kconfig and is not widely used (SMP affinity on multicore systems).

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: remove cmsis-rtos layering violation

We shouldn't be calling hooks from optional and upper layer subsystems
in the kernel, instead, just call the hook to set thread status in the
API where it is needed.

This now clears related bit in cmsis thread status bitarray when
terminating a thread in the cmsis rtos v1 layer directly and not in the
kenrel code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: move spinlock validation to own file

Move spin_lock validation outside of thread.c into own file. This code
really has nothing to do with the thread code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: split irq_offload ccode into own file

Move irq_offload code out of thread.c into own file. This code is not
related to threads.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: move essential flag related routines out

The functions to manipulate the essential flag indeed operate on
threads, but they are misplaced in the thread implementation file. Put
them alongside other routines setting other thread flags and cleanup
headers a bit.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: reorg Kconfigs and split them

The kernel kconfig is becoming too big and unmanageable with too many
options scattered across the file. Move some areas out and reorg main
Kconfig slightly.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: move z_init_static_threads to where it is being used

Move out of thread and put directly in init.c where it is being used.
Also remove definition from kernel.h, this is an internal function and
should not be in a public header.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

peter-mitsis commented on subsys/portability/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -197,6 +197,7 @@ osStatus osThreadTerminate(osThreadId thread_id)
 	}
 
 	k_thread_abort((k_tid_t)thread_id);
+	z_thread_cmsis_status_mask_clear((k_tid_t)thread_id);
```
1. If the thread self-terminates through osThreadTerminate, then the status mask will never be cleared. As a result, that thread will forever be seen as in use.
2. If there is a context switch between returning from `k_thread_abort()` and `z_thread_cmsis_status_mask_clear()`, the aborted thread will not be able to be reused until we context switch back to the original thread again.

I don't know enough about CMSIS to say for certain if either of the above scenarios would be an issue for CMSIS, but I wanted to bring it to attention in case that they might be.

nashif replied:
hmm, need to check, but add this line does the trick, without it, the cmsis v1 tests fail.

nashif replied:
I did it differently now

npitre commented:
You should s/SMP addinity/SMP affinity/ in commit log.

nashif committed:
kernel: do not export z_thread_priority_set

This function is only being used by a test, so instead of reimplementing
a syscall in the test, provide a Kconfig option to provide the
functionality that only works with tests and remove some of the
duplication and extra code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: sched: z_set_prio -> z_thread_prio_set

Rename private function to make it clear what priority we are setting
and to be consistent across the code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

cfriedt commented on kernel/Kconfig:
```diff
@@ -386,6 +349,46 @@ config WAITQ_DUMB
 
 endchoice # WAITQ_ALGORITHM
 
+menu "Misc Kernel related options"
```
This could probably be put into a `Kconfig.tls`

nashif committed:
kernel: split thread monitor

Move thread monitor related functions, not enabled in most cases outside
of thread.c and cleanup headers.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: remove unused code in headers

List of functions defined in headers and not being used anywhere.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
update:
- More cleanup commits

nashif committed:
kernel: move float operations out of thread.c

Move float operation out and add missing vrfy hook for enabling float.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: thread: use consistent thread parameter

Use thread wherever it makes sense, using 't' in some places can get
confused with 't' used for timeouts for example.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: rename thread/timeout parameter

Using t for both thread and timeout can be confusing at times.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: rename h -> heap

Avoid single characker variables that renders code unreadable and might
cause conflicts in maing, similar to t for both timeout and thread in
some places.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

peter-mitsis commented on kernel/Kconfig:
```diff
@@ -192,6 +192,11 @@ config THREAD_STACK_INFO
 	  the k_thread data structure.
 
+config THREAD_ABORT_HOOK
+	bool "Custom hook when aborting thread"
+	help
+	  Used by portability layers to modify locally managed status mask.
```
Please correct me if I am wrong ....

1. Doesn't the prompt suggest that this is user-facing?
2. If it is intended to be user-facing, then wouldn't we need to document the hook API and its signature?

As a potential alternative ... what if we used a weak function defined in sched.c and had a strong version of it defined in the CMSIS v1?

nashif replied:
> 1. Doesn't the prompt suggest that this is user-facing?

yes, removed

peter-mitsis commented on kernel/sched.c:
```diff
@@ -9,6 +9,7 @@
 #include <zephyr/kernel/internal/sched_priq.h>
 #include <wait_q.h>
 #include <kswap.h>
+#include <kthread.h>
```
I'm guessing that the inclusion of the kthread.h header file and the blank line below are artefacts of the development process and should likely be cleaned up.

peter-mitsis commented on kernel/Kconfig:
```diff
@@ -208,6 +208,13 @@ config THREAD_USERSPACE_LOCAL_DATA
+config USERSPACE_THREAD_MAY_RAISE_PRIORITY
+	bool "Thread can raise own priority"
+	depends on USERSPACE
+	depends on TEST # This should only be enabeld by tests.
```
Typo: enabeld -> enabled

nashif replied:
fixed

nashif commented on kernel/cpu_mask.c:
```diff
@@ -0,0 +1,76 @@
+#if defined(CONFIG_ASSERT) && defined(CONFIG_SCHED_CPU_MASK_PIN_ONLY)
+		int m = thread->base.cpu_mask;
+
+		__ASSERT((m == 0) || ((m & (m - 1)) == 0),
```
I am not changing or adding this code, lets fix this if needed in a followup PR.

nashif commented on kernel/cpu_mask.c:
```diff
@@ -0,0 +1,76 @@
+
+# ifdef CONFIG_SMP
+/* Right now we use a two byte for this mask */
+BUILD_ASSERT(CONFIG_MP_MAX_NUM_CPUS <= 16, "Too many CPUs for mask word");
```
correct, but an architecture can redefine this? like in arch/x86/core/Kconfig.intel64

you see:

```
config MP_MAX_NUM_CPUS
        range 1 4
```

This is lowering the number, but someone might also go over 12. Anyways, I am just moving code around, this can be dealt with in a followup

nashif commented on kernel/cpu_mask.c:
```diff
@@ -0,0 +1,76 @@
+#include <ksched.h>
+#include <zephyr/spinlock.h>
+
+extern struct k_spinlock sched_spinlock;
```
any suggestions? sched_spinlock seems fine because this is a scheduling spinlock afaik

peter-mitsis replied:
Possibly z_sched_spinlock? Or _sched_spinlock? I was just thinking along the lines of something to further decrease the likelihood that an application would choose to this name (thereby leading to linker trying and failing to choose the correct instance).

nashif committed:
spelling: add enabeld typo

add typo of enabeld.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: rename scheduler spinlock variable and make it private

rename sched_spinlock to _sched_spinglock to maintain it is privacy and
to avoid any misuse.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
